### ISSUE TITLE
ElasticSearch

### ISSUE THREAD

**Issue by CCoffie** [metabase/metabase#1300]
I'm not sure if metabase would be compatible with adding an ElasticSearch driver. If it is I would love to see it added.

**Comment by pking74** [comment]
I would like to see ElasticSearch supported as well.

**Comment by hilaryspencer** [comment]
Another vote for ES support

**Comment by yehosef** [comment]
+1, ES is amazing

**Comment by kodyo** [comment]
One more!

**Comment by decker502** [comment]
+1

**Comment by bvolders** [comment]
+1

**Comment by javadevmtl** [comment]
+1

**Comment by wheredevel** [comment]
+1 STRONG!

**Comment by mzimmerman** [comment]
Heard you on the changelog, +1 for me on elasticsearch!

**Comment by kaur2k** [comment]
+1 Would use it!

**Comment by tomferreira** [comment]
+1

**Comment by adrien-candiotti** [comment]
+1

**Comment by gincor** [comment]
+1

**Comment by spa-87** [comment]
+1

**Comment by kubbo** [comment]
+1

**Comment by eurotrip11** [comment]
+1

**Comment by edmarriner** [comment]
+1

**Comment by khalidsalomao** [comment]
+1

**Comment by jkarraker** [comment]
+1

**Comment by vikasgulati** [comment]
+1

**Comment by mariosergioti** [comment]
+1

**Comment by andreimavenhut** [comment]
+1

**Comment by vsreddy1602** [comment]
+1

**Comment by estewart9** [comment]
+1

**Comment by aksakal** [comment]
+1

**Comment by leomelzer** [comment]
+1

**Comment by alfredocambera** [comment]
+1

**Comment by juzerali** [comment]
👍 +1

**Comment by ptz0n** [comment]
+1

**Comment by coreation** [comment]
+1

**Comment by soaxelbrooke** [comment]
Grafana is doing "beautiful charts from elasticsearch" pretty well, it is a good place to take inspiration from:

http://docs.grafana.org/datasources/elasticsearch/

**Comment by tikh** [comment]
+1

**Comment by camsaul** [comment]
Hey everyone, one way we prioritize issues is by sorting them by the number of 👍 reactions. So instead of leaving `+1` comments, please just upvote by adding a reaction.

**Comment by ptz0n** [comment]
@camsaul You could also count the `+1`'s when prioritizing 😉

**Comment by camsaul** [comment]
@ptz0n, GitHub doesn't have a filter for counting `+1` comments. And it notifies everyone on a given issue when a new comment is added (in this case, 35 people). So please just use 👍  reactions.

**Comment by E14** [comment]
It does not really seem to matter though, as this is the top ticket anyway ;)

May I ask what parts pose a difficulty adding this as a source - I mean, is there a technical difficulty or just a time issue?

**Comment by fernandocchaves** [comment]
+1

**Comment by camsaul** [comment]
@fernandocchaves I mentioned this above but instead of leaving +1 or 👍 comments, PLEASE just upvote by adding a reaction to the issue description.

😒

**Comment by E14** [comment]
@camsaul 
> May I ask what parts pose a difficulty adding this as a source - I mean, is there a technical difficulty or just a time issue?

If it's just time and not complexity, I might just write the component myself, but writing a component that the maintainers already avoid due to complexity if probably not the best to start with -- that's why I ask.

**Comment by tlrobinson** [comment]
@E14 An Elasticsearch driver won't be as easy as adding a new SQL driver, but I don't believe it will be significantly more complex than some of our other drivers. If you want to take a look at the Druid, MongoDB, and Google Analytics drivers to see if you're up to the task that would be a good place to start. @camsaul correct me if I'm wrong.

Also note that in order to merge it into Metabase proper we'll need to make sure it's tested, including in our CI environment (CircleCI, see `circle.yml` and `bin/ci`)

**Comment by camsaul** [comment]
@E14 This is really more of a bandwidth thing than anything else -- we're a small team so we can only work on so many features at once. That being said, I'm not really sure how difficult this would be to implement. It's been probably 4 years since I've used ElasticSearch for anything so it's not really fresh in my memory right now. But I have a feeling it would be somewhat more difficult than other drivers like SQL-based ones or even Mongo since ElasticSearch's API is optimized more for search than the normal database operations we rely on for analysis.

Either way, you're welcome to give it a go. Check out our [Guide to Writing a Driver](https://github.com/metabase/metabase/wiki/Writing-A-Driver) for an overview on how they work.

**Comment by E14** [comment]
@tlrobinson / @camsaul Thank you for the reply and pointers! I'll have a look and see if it makes sense (for me ;)) to put time into it.

**Comment by hooopo** [comment]
+1

**Comment by E14** [comment]
Sorry for the missing feedback. For me there is too much documentation missing on the side of IDriver to start right into it without knowing the nuts&bolts and considering the time I have available right now.

What I can tell is that there is a [Java library available for ES](https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/index.html), so as with the Google drivers it is not necessary to implement the actual protocol, but only a wrapper around the library. Considering this, the driver might be better implemented as an optional add-in, especially since I have no idea under which license the client library can be used.

**Comment by globalundo** [comment]
For all the people waiting for this feature to be delivered - I can recommend to use a bit hacky approach for now: in our metabase installation, one of the data sources is *local postgresql database installation*, which had *multicorn* and *postgres_fwd* extensions installed. This way, you can access any other service as postgresql tables.
We are already using this to access data from Datadog and Mongodb installations, and plugins can be written in simple python - see https://github.com/globalundo/corndog or https://github.com/asya999/yam_fdw

**Comment by noblemavely** [comment]
+1

**Comment by Overbryd** [comment]
Please... stop posting "+1" to public isses.
Your message, and this reply, annoys all the people who are subscribed for
updates on this issue.

No one else cares if you "+1" this. If you want to express a vote on an
issue, use the "subscribe" to express interest or place a "thumbs up" emoij
on the respective issue.

**Comment by camsaul** [comment]
Seconding what @Overbryd said. One way we prioritize issues is by sorting them by the number of 👍 reactions on each issue's description. So instead of leaving `+1` or 👍  comments, please just upvote by adding a reaction to the issue description.

**Comment by porsovan** [comment]
elasticsearch is amazing

**Comment by je2ryw** [comment]
FYI, a WIP elasticsearch driver fork, and currently only support basic metadata discovery and native query with parameter variables. https://github.com/uxpsystems/metabase/tree/elasticsearch-driver

**Comment by fredpolicarpo** [comment]
+1

**Comment by prmdsharma-off** [comment]
+1

**Comment by allansene** [comment]
Elastic is experimenting with a SQL-like language. They just released a JDBC Driver.

I think that this may help:

https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-jdbc.html

**Comment by fetanchaud** [comment]
relates to #5562, as Dremio connects to ElasticSearch !

**Comment by vietvudanh** [comment]
Please add. ElasticSearch is fast and popular for analytics.

They are just REST request, JSON body so I don't think it would be too complicated. We can even use it might be just raw JSON and table like output (top level of JSON).

**Comment by bradjorgensen** [comment]
Also interested in an ElasticSearch driver.

**Comment by hodyroff** [comment]
The Amazon distribution of Elastic includes JDBC: https://opendistro.github.io/for-elasticsearch/downloads.html

**Comment by hadifarnoud** [comment]
It looks like this is not going to happen? it's been years since this issue created

**Comment by angjelo** [comment]
https://github.com/splitgraph/postgres-elasticsearch-fdw  .... I think if you use splitgraph and have  postgres you should be able to query elastic from there you can read and write too ...

**Comment by apinstein** [comment]
> The Amazon distribution of Elastic includes JDBC: https://opendistro.github.io/for-elasticsearch/downloads.html

Yes.. I just ran across this, too! That actually comes from the underlying ES project, and there is an OSS JDBC driver: https://github.com/opendistro-for-elasticsearch/sql

So I am correct in thinking all that needs to be done to support ES-SQL in metabase would be to create a new "Module" per https://github.com/metabase/metabase/wiki/Writing-a-Driver:-Packaging-a-Driver-&-Metabase-Plugin-Basics#drivers-shipped-as-3rd-party-plugins and is that still the best place to start?

**Comment by dacort** [comment]
> So I am correct in thinking all that needs to be done to support ES-SQL in metabase would be to create a new "Module" per https://github.com/metabase/metabase/wiki/Writing-a-Driver:-Packaging-a-Driver-&-Metabase-Plugin-Basics and is that still the best place to start?

Yep, that'd be the approach! I'd say go up one level in the wiki to https://github.com/metabase/metabase/wiki/Writing-A-Driver - that's got some high-level detail and links to other parts of the driver docs. The documentation is still a work in progress, but should get you started.

Also feel free to take a look at https://github.com/dacort/metabase-athena-driver for an example driver that's used in several production environments. The biggest challenge is probably just keeping the driver up-to-date with Metabase releases. We do occasionally make breaking changes (see the `Driver Notices` section on the wiki) but there's a mailing list you can subscribe to on https://www.metabase.com/docs/latest/developers-guide-drivers.html

**Comment by ArjonBu** [comment]
This indeed is a great feature request. What would make this happen? Would sponsorship speed this up? Maybe we all can donate something.

**Comment by apinstein** [comment]
Taking a stab at this today...

**Comment by apinstein** [comment]
Ok, I made some good progress! I actually have the driver getting configured properly and attempting to connect to the ESQL cluster.

I am getting an unexpected error from ES, though...

```
Server sent bad type [action_request_validation_exception]. Original type was [Validation Failed: 1: The [0.0.0] version of the [jdbc] client is not compatible with Elasticsearch version [7.10.1];
```

I cannot figure out how to specify the version of the JDBC client so that ES doesn't complain.

**Comment by apinstein** [comment]
First light!

<img width="1162" alt="image" src="https://user-images.githubusercontent.com/34889/103138453-c5ab9480-46a0-11eb-8727-3e1bc411acbd.png">

**Comment by TranDatDT** [comment]
Is there any update on this?

**Comment by ealexisaraujo** [comment]
> First light!
> 
> <img alt="image" width="1162" src="https://user-images.githubusercontent.com/34889/103138453-c5ab9480-46a0-11eb-8727-3e1bc411acbd.png">

Hi @apinstein could you please help me

We are working in elastic search in aws and I would like add the driver in metabase, do you have documentation and how to integrate this ? 

Thanks in advance.

**Comment by apinstein** [comment]
It is not working yet enough to be usable.

**Comment by ealexisaraujo** [comment]
> It is not working yet enough to be usable.

Okay, thanks for you answer

**Comment by romainquellec** [comment]
Hey ! Any update on this ?

**Comment by romainquellec** [comment]
Hey ! Any update on this ? Can't see it on the roadmap.

**Comment by carloserocha** [comment]
Hi everyone, any update about this?

**Comment by viniciusdsmello** [comment]
Hello everyone! Any updates on this topic?

**Comment by paoliniluis** [comment]
@viniciusdsmello use Trino in the meantime. There's a working demo here https://github.com/paoliniluis/metabase-trino

**Comment by canob** [comment]
> @viniciusdsmello use Trino in the meantime. There's a working demo here https://github.com/paoliniluis/metabase-trino

Trino is a good option, but is really slower compared to for example use the JDBC of ElasticSearch or OpenSearch (we are talking about seconds vs minutes in the query search time), I already tried this some time ago and it was not enough

**Comment by viniciusdsmello** [comment]
Thank you, @paoliniluis and @canob! I'll try it :)

**Comment by grapergrape** [comment]
Any update? this has been in the pipeline for almost 10 years + minutes to seconds difference in using trino is a dealbreaker

**Comment by canob** [comment]
> @canob @grapergrape have you contacted the Trino team to see why there's so much difference?

Hi Luis,

I think the difference is because the underlying technology that Trino uses. Trino is slower to connect to any data source compared to connect to them directly, but provide advanced functions, like joins between different types of databases engines.

The best option here for get more speed is to have a native driver for ElasticSearch/OpenSearch in Metabase. You can increase the speed of Trino for example using clustering, but is not simple.

**Comment by sibelius** [comment]
is there any open source workaround for this?

**Comment by paoliniluis** [comment]
@sibelius use trino

**Comment by rserit** [comment]
Any update? This is very much needed.

**Comment by yavuzyazici** [comment]
> Any update? This is very much needed.

+1

**Comment by mhkarimi1383** [comment]
+1

**Comment by yavuzyazici** [comment]
We created a driver on our own as Atak Domain team

https://github.com/atakdomain/metabase-elasticsearch-driver

it works pretty fine for Raw DSL native queries and SQL alike searchs

examples:
SELECT * FROM "something_domain_something"
	WHERE "objectName.keyword" = 'atakdomain.com' AND "executedMethod" != 'GetInfoAsync'
	ORDER BY requestDate DESC
LIMIT 1000

GET something_domain_something/_search
{
  "size": 200, 
  "query": {
    "bool": {
      "must": [
        {
          "match_phrase": {
            "objectName": "atakdomain.com"
          }
        }
      ]
    }
  },
  "sort": [
    {
      "requestDate": {
        "order": "desc"
      }
    }
  ]
}

**Comment by sibelius** [comment]
Trino
